### PR TITLE
PF-423: Separate exceptions into user-facing and internal.

### PR DESCRIPTION
**Pull request by marikomedlock**
- Separated exceptions into user-actionable and system error sub-classes. Replaced all RuntimeExceptions with one of the new sub-classes.
- Added a custom exception handler to the main `picocli` class to handle user-actionable and system, and unexpected exceptions differently.
- Added two logback appenders: console and file (`$HOME/.terra/terra.log` in the global context directory). Replaced the `logback.xml` file with programmatic logging setup, so that we can manipulate the log level via a command in a future PR (see [PF-542](https://broadworkbench.atlassian.net/browse/PF-542)).
- The goal of the two appenders is to have more verbose logging go to the file, and shorter error messages go to the console. To this end, the default log levels are `OFF` and `INFO` for console and file, respectively.
- The file appender rolls log files. It stores a maximum of 5 log files at a time, of at most 5 MB each. That seems way more than enough for the usage patterns I'd expect for this CLI, so we may want to adjust that down in the future.
- Explicitly specified a color scheme for the commands. User actionable errors, including mistyping commands, prints in regular blue text. System and unexpected errors print with a bold red [ERROR] in front of the actual message.

Note to reviewers: There are a lot of files changed, but most are just replacements of `RuntimeException` with one of the new exception sub-classes, or changes to `logger.__` calls to make the levels more consistent across classes. The main things to review are:
- `Main.java` is the CLI entry point (i.e. the Java `main` method) and contains the custom exception handler.
- `SystemException.java` and `UserActionableException.java` are the new exception sub-classes.
- `GlobalContext.java` contains the pointers to the log levels and log file.
- `Logger.java` contains the appenders setup.

### CODE REVIEW

File: src/main/java/bio/terra/cli/command/Main.java
Comment (jmczerk) [inline src/main/java/bio/terra/cli/command/Main.java:102]: It looks like `Logger#LOG_FORMAT` is configured to prefix messages with their log level?  If so, you might consider moving the `"[ERROR] "` prefix append to `printErrorText()` time.

File: src/main/java/bio/terra/cli/command/Main.java
Comment (jmczerk) [inline src/main/java/bio/terra/cli/command/Main.java:109]: Nit: could consider moving the ` + ": "` into the ternary true condition to prevent trailing colon when `getCause()` is null.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (jmczerk) [inline src/main/java/bio/terra/cli/context/utils/Logger.java:41]: Is rolling the logs a concern for the CLI, or do we expect them to be pretty self-limiting size-wise due to the usage model?

**Review by jmczerk — APPROVED** [review]
A couple minor questions/nits, but looks good otherwise.

File: src/main/java/bio/terra/cli/command/Main.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/Main.java:102]: Yes, you're right. Done.

File: src/main/java/bio/terra/cli/command/Main.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/Main.java:109]: Agreed, done.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/utils/Logger.java:41]: Good point. I think they'll be pretty self-limiting due to usage, but it's probably better to roll the files just in case. I'll take a stab at it.

File: src/main/java/bio/terra/cli/auth/AuthenticationManager.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/auth/AuthenticationManager.java:207]: [follow-up] Not for this PR, but as we look towards an external-facing release, it will make sense to scrub through these exceptions for acronyms or words that won't make sense to end users. For internal system errors, where we don't think they're user caused or user actionable, we may want to err on the side of more opaqueness in the user-facing message. (Of course there's a limit... I wouldn't just give a made-up error code, but possibly a generic message like "An error occurred while fetching Terra service credentials for your user account" combined with a more precise error locator code "(code 1234)".

File: src/main/java/bio/terra/cli/command/Main.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/command/Main.java:104]: Potential formatting tweak to give users a bit more context:

"An unexpected error occurred in {className}: {message}"

File: src/main/java/bio/terra/cli/command/Main.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/command/Main.java:80]: (Caveat: I may be dramatically over-thinking things here. Take with a grain of salt and feel free to ignore at your leisure!)

I thought I was grokking the categorization here until about 2/3 way through this PR, when I realized that both user-facing and Internal exceptions are getting shown to the user with very similar formatting... so now I'm having more trouble precisely distinguishing between the two use cases.

The main difference in handling seems to be that (1) for Internal exceptions we'll include a message from the causal exception, and (2) for Internal exceptions we point users to the log file for more details (presumably so they can find a stack trace).

But for both types of exceptions, we're showing users a big red '[ERROR]'-prefixed line with the exception message included. This feels quite similar to me, and may make it hard to describe clear conventions around what level of user-facing polish should be applied to each type of exception.

Spitballing a couple potential changes that may help draw out more of a difference:

1) Consider naming options.

Surveying the usage patterns, `UserFacingException` is mostly thrown when there is a simple precondition failure (e.g. "no workspace context") or a par-for-the-course error code returned from the service (e.g. "resource not found"). Maybe these could be called `UserActionableException` or simply `UserException`? Whereas `InternalErrorException` is usually about a server failure (5xx), file parsing failure, or some other internal workings of the system. Possibly `SystemException` captures that concisely? I realize this isn't too different from the current names... but somehow I found the `UserFacing` term slightly jarring, because ultimately all of these exceptions are shown to the user.

2) Consider a formatting difference between user-actionable and internal system exceptions.

If I use incorrect characters in an object name or try to fetch an ID that doesn't exist, I might expect a less-scary looking error message than if a server is down or there is an error parsing some core config file. To some extent, for the former type of "user error", I'd probably want it to look similar to what happens when I mis-type a command name. So maybe there's some stronger delineation we can aim for here in the formatted output, e.g.

UserActionable: normal color, less loud (maybe drop the [ERROR] prefix?), looks like a simple message "Hey, you did something wrong" or "hey, we couldn't find that thing you asked for".
Internal / system: some red somewhere (maybe just the [ERROR] prefix is red), with loud with clear boilerplate indicating that a relatively serious system error has occurred.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/context/utils/Logger.java:23]: nit: this is more than just utility methods. Maybe re-phrase as "Utility class to initialize a Logback context for the main CLI app. Requires an already-loaded GlobalContext for defining logging file outputs and loading config properties."

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/context/utils/Logger.java:146]: nit / side note: I'm a little surprised there isn't some built-in class that does this already.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/context/utils/Logger.java:158]: [No action / side comment] I *think* I understand the intent here. But the only use case I can envision for this is in GlobalContext to configure the logback log level (either as part of a stored configuration or via command-line arguments). In which case, it's not too inappropriate to say that GlobalContext should know that our implementation is logback, because it's directly configuring the logback system.

Is there some other scenario you're protecting against, where these levels become more widely-used throughout the codebase?

Or, maybe the concern is more about how these values might get serialized alongside the GlobalContext? In which case a small note about serialization control might help clarify the need.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/context/utils/Logger.java:52]: [Side note / comment] While it might have technically been possible to set up a console / file-based output with a logback.xml and various properties getting passed in (like the location of the log file), 💯 to doing things programmatically — it's far easier to understand, debug, and improve the flow over time. I ran into a similar challenge in setting up JSON logging on the service side, with a similar solution. (In fact there are some improvements I'd like to make after reading your code here, like the `StatusPrinter.printIfErrorsOccurred` safety net.)

**Review by gjuggler — APPROVED** [review]
Nice improvements! I dropped a bunch of fairly long comments / ideas, but there's nothing that I would block merging on. If you do pick up any larger changes feel free to ping me for another look, but I'm flipping the approval bit now.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (jmczerk) [inline src/main/java/bio/terra/cli/context/utils/Logger.java:58]: Thanks for adding the `RollingFileAppender`!

File: src/main/java/bio/terra/cli/command/exception/UserFacingException.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/exception/UserFacingException.java:5]: Good point. Added.

File: src/main/java/bio/terra/cli/context/GlobalContext.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/GlobalContext.java:64]: This is a pretty common error case (e.g. file not found when a user first runs the CLI and there is no pre-existing global context file) and it should not be fatal. We handle this by returning an object populated with default values in the lines below. So, I think there's no need to log or print the error returned here. Added a comment to this effect in the catch block.

File: src/main/java/bio/terra/cli/command/Main.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/Main.java:99]: This looks great. Agreed this is probably a couple months away, but I'll keep in mind for then. Thanks!

File: src/main/java/bio/terra/cli/auth/AuthenticationManager.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/auth/AuthenticationManager.java:191]: Added.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/utils/Logger.java:58]: Thank you for the great suggestion! I'm working on addressing Greg's comments now, and then I think I'm going to ping you both for a re-review, if you don't mind. Enough has changed since the original that another once-over by you guys would be much appreciated.

File: src/main/java/bio/terra/cli/auth/AuthenticationManager.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/auth/AuthenticationManager.java:207]: Good idea. Added this to the list of style improvements in [PF-574](https://broadworkbench.atlassian.net/browse/PF-574).

File: src/main/java/bio/terra/cli/command/Main.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/Main.java:104]: Changed.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/utils/Logger.java:23]: Updated.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/utils/Logger.java:146]: You're right. Found `ThresholdFilter` and used that instead.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/utils/Logger.java:158]: Without this wrapper class, we'd leak logger implementation into:
- GlobalContext, including the serialized version persisted in a file.  ch.qos.logback.classic.Level is an object, not a string, and I didn't want this file to include logger implementation-specific object structure. That would make it harder to change later (i.e. because if we replace the cho.qos library with another one, then we won't have the appropriate  Level class any more to do the deserialization)
- (Upcoming) Config command to set the log level(s), including displaying the possible values.

Added a note about this to the code.

File: src/main/java/bio/terra/cli/command/Main.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/Main.java:80]: This makes sense about there not being a big enough difference between user-actionable and system/unexpected exceptions.
1. Renamed the exception classes to UserActionableException and SystemException.
2. Changed the formatting so that system and unexpected exceptions show a bold red message, and user-actionable exceptions show a regular size blue message. Since command typing errors (e.g. forgot a required flag) are also user-actionable, I also changed them to follow this convention.

I also explicitly defined the color scheme in the `Main` class, so that we can modify it more easily in the future. Below is an example of what it looks like now (I inserted test exceptions in the `terra status` and `terra version` commands):
Screen Shot 2021-03-17 at 2 20 39 PM

Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/Main.java]: I don't have any strong preferences around the colors here, so if there's any changes you guys would like to see, just let me know. (There's no orange though FYI 🤷‍♀️ ).

**Review by gjuggler — APPROVED** [review]
🥳  awesome and thoughtful changes! And thanks for the screenshot. That looks really polished, IMO.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (jmczerk) [inline src/main/java/bio/terra/cli/context/utils/Logger.java:29]: Intuitively 50 MB * 5 files seems an order of magnitude high to me; maybe 5MB * 5?  I'll defer to your intuition though, as I don't know how big these get when debugging...

**Review by jmczerk — APPROVED** [review]
Thanks for the screenshot, it was really helpful to see the color scheme in action.  One minor note on log file limits, but otherwise this looks great to me.

File: src/main/java/bio/terra/cli/context/utils/Logger.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/utils/Logger.java:29]: I have no idea either. We may want to change this in the future once we actually start using these debug logs. Mostly I'm thinking of these logs as a hedge for when people other than us start installing this. Changed to 5MB * 5 instead.

**Pull request merged**
